packages/client/hmi-client/src/services/calibrate-workflow.ts

```diff
@@ -117,76 +127,172 @@ export const renderLossGraph = (
-	if (!allSimilarity) return result;
-	// normalized cosine similarities
-	const filteredSim = allSimilarity.filter((ele) => ele.similarity > acceptableDistance);
-	filteredSim.forEach((sim) => {
-		// Find all states assosiated with this sim
-		const validStates = modelOptions.filter((state) => {
-			if (state.grounding?.identifiers) {
-				const modelGroundingList = Object.entries(state.grounding?.identifiers).map((ele) =>
-					ele.join(':')
-				);
-				return modelGroundingList.includes(sim.source);
-			}
-			return false;
-		});
-		// Find all columns assosiated with this sim
-		const validCols = datasetOptions.filter((col) => {
-			if (col.metadata?.groundings?.identifiers) {
-				const dataGroundingList = Object.entries(col.metadata?.groundings?.identifiers).map((ele) =>
-					ele.join(':')
-				);
-				return dataGroundingList.includes(sim.target);
-			}
-			return false;
-		});
-		// For all states and columns that have short distances throw them into results
-		validStates.forEach((state) => {
-			validCols.forEach((col) => {
-				result.push({ modelVariable: state.id, datasetVariable: col.name });
+	const entityResult = await autoEntityMapping(sourceEntities, targetEntities, acceptableDistance);
+
+	// rename result to CalibrateMap for users of this function
+	entityResult.forEach((entity) => {
+		result.push({ modelVariable: entity.source, datasetVariable: entity.target });
+	});
+	return result;
+};
+
+// Takes in two lists of states
+// Transforms them into generic entities with {id, groundings}
+// Uses autoEntityMapping to determine 1:1 mapping
+// rewrites result in form {modelOneVariable, modelTwoVariable}
+export const autoModelMapping = async (modelOneOptions: State[], modelTwoOptions: State[]) => {
+	const result = [] as any[];
+	const acceptableDistance = 0.5;
+	const sourceEntities: Entity[] = [];
+	const targetEntities: Entity[] = [];
+
+	// Fill sourceEntities with modelOneOptions
+	modelOneOptions.forEach((state) => {
+		if (state.grounding?.identifiers) {
+			sourceEntities.push({
+				id: state.id,
+				groundings: Object.entries(state.grounding?.identifiers).map((ele) => ele.join(':'))
 			});
-		});
+		} else {
+			sourceEntities.push({ id: state.id });
+		}
 	});
 
-	// due to a state and a column having potential for multiple pairwise matches, lets remove duplicates from results.
-	const distinctResults = result.filter(
-		(value, index) =>
-			index ===
-			result.findIndex(
-				(obj) =>
-					obj.datasetVariable === value.datasetVariable && obj.modelVariable === value.modelVariable
-			)
-	);
+	// Fill targetEntities with datasetOptions
+	modelTwoOptions.forEach((state) => {
+		if (state.grounding?.identifiers) {
+			sourceEntities.push({
+				id: state.id,
+				groundings: Object.entries(state.grounding?.identifiers).map((ele) => ele.join(':'))
+			});
+		} else {
+			sourceEntities.push({ id: state.id });
+		}
+	});
```

Comment:
Same pattern as above. Not sure if you want to make a helper function for this since it's used quite a bit
```suggestion
	// Fill sourceEntities with modelOneOptions
	modelOneOptions.forEach((state) => {
		const groundings = state.grounding?.identifiers
			? Object.entries(state.grounding.identifiers).map((ele) => ele.join(':'))
			: undefined;
		sourceEntities.push({ id: state.id, groundings });
	});

	// Fill targetEntities with datasetOptions
	modelTwoOptions.forEach((state) => {
		const groundings = state.grounding?.identifiers
			? Object.entries(state.grounding.identifiers).map((ele) => ele.join(':'))
			: undefined;
		targetEntities.push({ id: state.id, groundings });
	});
```

Reply:
I find this pattern harder to see explicitly what is happening but you do make a good point that the way I have written it its harder to see that `sourceEntities` is being pushed to either way
I will take your approach here